README.md
# Status
It works! v0.1.0 released 11 Nov. 2019!

# PDF2SearchablePDF
`pdf2searchablepdf <input.pdf>` = voila! "input_searchable.pdf" has now been created and has searchable text!

# Description:
`tesseract` has the ability to do OCR (Optical Character Recognition) on image files, but unfortunately NOT on PDF files as inputs. This is unfortunate, as it means it's a pain to try to convert a PDF to a searchable PDF, so this is an attempt at scripting the process using existing tools in order to make it stupid-simple for ANYONE to use!

# Dependencies:
This has been tested only on Ubuntu 18. It requires the following programs:

## You Must Install these:

    sudo apt update 
    sudo apt install tesseract-ocr

See: https://github.com/tesseract-ocr/tesseract/wiki

## It also relies on these, but they come pre-installed on Ubuntu 18:

1. `pdftoppm`

# Installation
Simply run the "install.sh" script to create a symbolic link to `pdf2searchablepdf` in your "~/bin" directory:

	./install.sh

# Sample run and output:

```
$ pdf2searchablepdf ./test_pdfs/test1.pdf 
pdf2searchablepdf version 0.1.0
=================================================================================
Converting input PDF (./test_pdfs/test1.pdf) into a searchable PDF
=================================================================================
Creating temporary working directory: "pdf2searchablepdf_temp_20191111-001431.509915114"
Converting input PDF to a bunch of output TIF images inside temporary working directory.
- THIS COULD TAKE A LONG TIME (up to 45 sec or so per page)! Manually watch the temporary
  working directory to see the pages created one-by-one to roughly monitor progress.
- NB: each TIF file created is ~25MB, so ensure you have enough disk space for this
  operation to complete successfully.
All TIF files created.
Running tesseract OCR on all generated TIF images in the temporary working directory.
This could take some time.
Searchable PDF will be generated at "./test_pdfs/test1_searchable.pdf".
Tesseract Open Source OCR Engine v4.0.0-beta.1 with Leptonica
Page 0 : pdf2searchablepdf_temp_20191111-001431.509915114/pg-1.tif
Page 1 : pdf2searchablepdf_temp_20191111-001431.509915114/pg-2.tif
Page 2 : pdf2searchablepdf_temp_20191111-001431.509915114/pg-3.tif
Done! Searchable PDF generated at "./test_pdfs/test1_searchable.pdf".
Removing temporary working directory at "pdf2searchablepdf_temp_20191111-001431.509915114".
Done!

Total script run-time: 136 sec
END OF pdf2searchablepdf.

```


changelog.md

# Changelog
- Newest on top
- Follows Semantic Versioning; see: https://semver.org/

In short: 
Given a version number MAJOR.MINOR.PATCH, increment the:  

MAJOR version when you make incompatible API changes,  
MINOR version when you add functionality in a backwards compatible manner, and  
PATCH version when you make backwards compatible bug fixes.  


## 2019-11-10 - v0.1.0
Initial release. It works!

install.sh
# "Install" this script by symbolically linking from here to your "~/bin" dir! 
# - NB: don't delete your files here or it will break the "install", since it's simply symbolically linking to 
#   the executable here! 
# - Therefore, if you move these files, simply re-run this install script and it will automatically update the 
#   symbolic link in ~/bin and all will be well again! 
# - Note: this script does NOT add the "~/bin" dir to your PATH. Run `echo $PATH` and ensure you see an entry like 
#   this: `/home/my_username/bin:`. If you don't, you'll need to manually add your "~/bin" directory to your path
#   for this program to be visible.

mkdir -p ~/bin

# Obtain full path to this install.sh file:
full_path=$(readlink -f $0)
# echo "full_path = $full_path"

# Get just the directory path now, by stripping off the filename part; see: https://stackoverflow.com/a/6121114/4561887
dir=$(dirname "${full_path}")
# echo "dir = $dir"

path2exec="${dir}/pdf2searchablepdf.sh" # path to the executable bash script
# echo "path2exec = $path2exec"

ln -sf $path2exec ~/bin/pdf2searchablepdf

echo "Done: symbolic link should have been placed in \"~/bin/pdf2searchablepdf\"."

pdf2searchablepdf.sh
#!/bin/bash

# pdf2searchablepdf
# Gabriel Staples
# Started: 9 Nov. 2019 
# - Source code: https://github.com/ElectricRCAircraftGuy/PDF2SearchablePDF

# Test run: `pdf2searchablepdf ./test_pdfs/test1.pdf`

start=$SECONDS

EXIT_SUCCESS=0
EXIT_ERROR=1

VERSION="0.1.0"
AUTHOR="Gabriel Staples"

print_help() {
	echo "Purpose: convert \"input.pdf\" to a searchable PDF named \"input_searchable.pdf\""
	echo "by using tesseract to perform OCR (Optical Character Recognition) on the PDF."
	echo 'Usage: `pdf2searchablepdf <input.pdf>`'
	echo "Source code: https://github.com/ElectricRCAircraftGuy/PDF2SearchablePDF"
}

print_version() {
	echo "pdf2searchablepdf version $VERSION"
}

if [ $# -eq 0 ]; then
	echo "No arguments supplied"
	print_help
	exit $EXIT_ERROR
fi

# Help menu
if [ "$1" == "-h" ]; then
	print_help
	exit $EXIT_SUCCESS
fi

# Version
if [ "$1" == "-v" ]; then
	print_version
	echo "Author = $AUTHOR"
	echo 'See `pdf2searchablepdf -h` for more info.'
	exit $EXIT_SUCCESS
fi

pdf_in=$1
# Strip file extension; see: https://stackoverflow.com/a/32584935/4561887
pdf_in_no_ext=$(echo $pdf_in | rev | cut -f 2- -d '.' | rev)
pdf_out="${pdf_in_no_ext}_searchable"

print_version
echo "================================================================================="
echo "Converting input PDF ($pdf_in) into a searchable PDF"
echo "================================================================================="

# 1. First, create temporary directory to place all intermediate files
# - name it "pdf2searchablepdf_temp_yyyymmdd-hhmmss.ns"

timestamp=$(date '+%Y%m%d-%H%M%S.%N')
# echo "timestamp = $timestamp"
temp_dir="pdf2searchablepdf_temp_${timestamp}"
# echo "temp_dir = $temp_dir"
echo "Creating temporary working directory: \"$temp_dir\""
mkdir -p $temp_dir
rm -rf $temp_dir/* # ensure it is empty

# 2. Convert the input pdf to a bunch of TIF files inside this directory
# - See my ans here: https://askubuntu.com/questions/150100/extracting-embedded-images-from-a-pdf/1187844#1187844
echo "Converting input PDF to a bunch of output TIF images inside temporary working directory."
echo "- THIS COULD TAKE A LONG TIME (up to 45 sec or so per page)! Manually watch the temporary"
echo "  working directory to see the pages created one-by-one to roughly monitor progress."
echo "- NB: each TIF file created is ~25MB, so ensure you have enough disk space for this"
echo "  operation to complete successfully."
# TODO: fix this in the future to print out a progress bar in the form of 1 dot every 10 sec, and a % complete
# every time another file is completed, based on how many total files there are in the pdf! Be sure to line wrap 
# ever 80 chars or so as well. This will require spinning off another process in the background that does a file
# check once per second or so to monitor progress. Once all files are created I can kill the background process
# which was doing that monitoring.
pdftoppm -tiff -r 300 $pdf_in $temp_dir/pg

# FOR DEVELOPMENT TO SPEED THIS UP BY USING PREVIOUSLY-GENERATED FILES INSTEAD 
# (comment out the above command, & uncomment the below command):
# cp pdf2searchablepdf_temp_20191110-231200.594322352/* $temp_dir

echo "All TIF files created."

# 3. Create a text file containing a list of all of the generated tif files
# - Use "version sort", or `sort -V` to enforce proper sorting between numbers which are multiple digits 
# vs 1 digit--ex: "pg-1.tiff" and "pg-10.tiff", for instance. See here: https://unix.stackexchange.com/a/41659/114401
find $temp_dir/* | sort -V > $temp_dir/file_list.txt

# 4. Run tesseract OCR on all generated TIF images.
# See: https://github.com/tesseract-ocr/tesseract/wiki/FAQ
echo "Running tesseract OCR on all generated TIF images in the temporary working directory."
echo "This could take some time."
echo "Searchable PDF will be generated at \"${pdf_out}.pdf\"."
tesseract $temp_dir/file_list.txt $pdf_out pdf
echo "Done! Searchable PDF generated at \"${pdf_out}.pdf\"."

# 5. Delete temp dir
echo "Removing temporary working directory at \"$temp_dir\"."
rm -rf $temp_dir
echo "Done!"


end=$SECONDS
duration_sec=$(( end - start ))
echo -e "\nTotal script run-time: $duration_sec sec"

echo "END OF pdf2searchablepdf."
